pkg/virtualgarden/kube_api_server_certificates.go

```diff
@@ -0,0 +1,258 @@
+	if err != nil {
+		return err
+	}
+
+	_, err = o.deployKubeApiServerClientAdminCertificate(ctx, apiServerCACertificate, loadbalancer, checksums)
+	if err != nil {
+		return err
+	}
+
+	_, err = o.deployKubeApiServerMetricsScraperCertificate(ctx, apiServerCACertificate, checksums)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *operation) deleteKubeAPIServerCertificates(ctx context.Context) error {
+	o.log.Infof("Deleting secrets containing kube-apiserver certificates")
+
+	for _, name := range []string{
+		KubeApiServerSecretNameAggregatorCACertificate,
+		KubeApiServerSecretNameAggregatorClientCertificate,
+		KubeApiServerSecretNameApiServerCACertificate,
+		KubeApiServerSecretNameApiServerServerCertificate,
+		KubeApiServerSecretNameKubeControllerManagerCertificate,
+		KubeApiServerSecretNameClientAdminCertificate,
+		KubeApiServerSecretNameMetricsScraperCertificate,
+	} {
+		secret := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: o.namespace}}
+		if err := o.client.Delete(ctx, secret); client.IgnoreNotFound(err) != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *operation) deployKubeApiServerApiServerCACertificate(ctx context.Context, checksums map[string]string) (*secretsutil.Certificate, error) {
+	certConfig := &secretsutil.CertificateSecretConfig{
+		Name:       KubeApiServerSecretNameApiServerCACertificate,
+		CertType:   secretsutil.CACert,
+		CommonName: Prefix + ":ca:kube-apiserver",
+	}
+
+	cert, checksum, _, err := o.deployCertificate(ctx, certConfig, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	o.exports.KubeApiserverCaPem = string(cert.CertificatePEM)
+
+	checksums[ChecksumKeyKubeAPIServerCA] = checksum
+	return cert, err
+}
+
+func (o *operation) deployKubeApiServerApiServerServerCertificate(ctx context.Context, caCertificate *secretsutil.Certificate,
+	loadbalancer string, checksums map[string]string) (*secretsutil.Certificate, error) {
+	dnsAccessDomain := o.imports.VirtualGarden.KubeAPIServer.DnsAccessDomain
+
+	certConfig := &secretsutil.CertificateSecretConfig{
+		Name:       KubeApiServerSecretNameApiServerServerCertificate,
+		CertType:   secretsutil.ServerCert,
+		SigningCA:  caCertificate,
+		CommonName: Prefix + ":server:kube-apiserver",
+		DNSNames: []string{
+			"localhost",
+			KubeAPIServerServiceName,
+			KubeAPIServerServiceName + ".garden",
+			KubeAPIServerServiceName + ".garden.svc",
+			KubeAPIServerServiceName + ".garden.svc.cluster",
+			KubeAPIServerServiceName + ".garden.svc.cluster.local",
+			"kubernetes",
+			"kubernetes.default",
+			"kubernetes.default.svc",
+			"kubernetes.default.svc.cluster",
+			"kubernetes.default.svc.cluster.local",
+			fmt.Sprintf("api.%s", dnsAccessDomain),
+			fmt.Sprintf("gardener.%s", dnsAccessDomain),
+		},
+		IPAddresses: []net.IP{
+			net.ParseIP("127.0.0.1"),
+			net.ParseIP("100.64.0.1"),
+			net.ParseIP(loadbalancer),
+		},
+	}
```

Comment:
This fails when
  -  dnsAccessDomain is not set
  - loadbalancer uses a dns domain, not an IP (e,g ELB)

Use this instead:

```
loadbalancer string, checksums map[string]string) (*secretsutil.Certificate, error) {

	certConfig := &secretsutil.CertificateSecretConfig{
		Name:       KubeApiServerSecretNameApiServerServerCertificate,
		CertType:   secretsutil.ServerCert,
		SigningCA:  caCertificate,
		CommonName: Prefix + ":server:kube-apiserver",
		DNSNames: []string{
			"localhost",
			KubeAPIServerServiceName,
			KubeAPIServerServiceName + ".garden",
			KubeAPIServerServiceName + ".garden.svc",
			KubeAPIServerServiceName + ".garden.svc.cluster",
			KubeAPIServerServiceName + ".garden.svc.cluster.local",
			"kubernetes",
			"kubernetes.default",
			"kubernetes.default.svc",
			"kubernetes.default.svc.cluster",
			"kubernetes.default.svc.cluster.local",
		},
		IPAddresses: []net.IP{
			net.ParseIP("127.0.0.1"),
			net.ParseIP("100.64.0.1"),
		},
	}

	dnsAccessDomain := o.imports.VirtualGarden.KubeAPIServer.DnsAccessDomain
	if len(dnsAccessDomain) > 0 {
		certConfig.DNSNames = append(certConfig.DNSNames, fmt.Sprintf("api.%s", dnsAccessDomain), fmt.Sprintf("gardener.%s", dnsAccessDomain))
	}

	loadbalancerIP := net.ParseIP(loadbalancer)
	if loadbalancerIP != nil && len(loadbalancerIP.String()) != 0 {
		certConfig.IPAddresses = append(certConfig.IPAddresses, loadbalancerIP)
	} else {
		// because it is not an ip address, we assume it is a DNS domain
		certConfig.DNSNames = append(certConfig.DNSNames, loadbalancer)
	}
```

Reply:
A very good point. Thanks. We have changed this as you suggested.